File: 401-500/471-480/472-concatenatedWords/concatenatedWords.java
class Solution {
    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        List<String> ans = new ArrayList<>();
        HashSet<String> wordSet = new HashSet<>(Arrays.asList(words));

        for (String word : words) if (dfs(word, wordSet))
            ans.add(word);
            
        return ans;
    }

    boolean dfs(String word, HashSet<String> wordSet) {
        for (int i = 1; i < word.length(); i++) {
            if (wordSet.contains(word.substring(0, i))) {
                String suffix = word.substring(i);
                if (wordSet.contains(suffix) || dfs(suffix, wordSet)) {
                    wordSet.add(word); // can treat concatenated word as a new word for quickly lookup later
                    return true;
                }
            }
        }

        return false;
    }
}

File: 401-500/471-480/472-concatenatedWords/concatenatedWords.py
class Solution(object):
    def findAllConcatenatedWordsInADict(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """

        words.sort(key=lambda x: len(x))
        pre_word = set([words[0]])
        result = []

        for i in range(1, len(words)):
            word = words[i]
            length = len(word)
            dp = [True] + [False] * length

            for j in range(1, length+1):
                for k in range(j):
                    if dp[k] and word[k:j] in pre_word:
                        dp[j] = True
                        break

            if dp[-1]:
                result.append(word)

            pre_word.add(word)


        return result